File: src/day15.exs
import Enum
defmodule Navigate do
  defstruct grid: "", width: 0, height: 0
  
  @moves [{1,0},{0,1},{-1,0},{0,-1}]

  def load_file(path, ex \\ 0, ey \\ 0) do
    StringIO.open("", fn pid ->
      {width, height} = 0..ey |> reduce({0,0}, fn sy, {_width, _height} ->
        File.stream!(path) |> reduce({0,0}, fn line, {_w,h} ->
          line = String.trim(line)
          width = byte_size(line)
          costs = line |> :binary.bin_to_list |> map(&(&1 - 48))
          0..ex |> each(fn sx ->
            bytes = costs |> map(fn cost ->
              case rem(cost + sx + sy, 9) do
                0 -> 9
                c -> c
              end
            end) |> :binary.list_to_bin
            IO.write pid, bytes
          end)
          {width, h + 1}
        end)
      end)
      {_, grid} = StringIO.contents pid
      %Navigate{ grid: grid, width: width * (ex + 1), height: height * (ey + 1) }
    end) |> elem(1)
  end
  
  def cost(nav, start, goal), do: costr({[{heuristic_cost(start,goal), start}], %{start => 0}}, nav, goal)
  
  defp costr({[{_, node} | _], costs}, _nav, goal) when node == goal, do: Map.get(costs, node)

  defp costr({[{_, node} | frontier], costs}, nav, goal) do
    cost_so_far = Map.get(costs, node)
    neighbors(nav, node) |> reduce({frontier, costs}, fn next, {frontier, costs} ->
      with cost <- (node_cost(nav, next) + cost_so_far),
           {:ok, costs} <- new_best(costs, next, cost) do
        {enqueue(frontier, {heuristic_cost(next, goal) + cost, next}), costs}
      else
        _ -> {frontier, costs}
      end
    end) |> costr(nav, goal)
  end

  defp enqueue(list, item) do
    {cost, _} = item
    {left, right} = split_while(list, fn {lcost, _} -> lcost < cost end)
    left ++ [item | right]
  end

  def heuristic_cost({ax,ay}, {bx,by}) do
    abs(bx-ax) + abs(by-ay)
  end

  def node_cost(nav, {x,y}) do
    index = y * nav.width + x
    :binary.at(nav.grid, index)
  end

  def neighbors(nav, {x,y}) do
    @moves |> flat_map(fn {dx,dy} ->
      x = x + dx
      y = y + dy
      if x in 0..(nav.width-1) and y in 0..(nav.height-1) do
        [{x,y}]
      else
        []
      end
    end)
  end

  defp new_best(known_costs, node, cost) do
    Map.get_and_update(known_costs, node, fn
      v when cost < v -> {:ok, cost}
      v -> {:err, v}
    end)
  end
end

:timer.tc(fn ->
  Navigate.load_file("input/day15",4,4)
  |> Navigate.cost({0,0},{499,499})
end) |> IO.inspect

File: src/day15.rb
class Nav
  def self.parse(filename)
    qgrid = File.read(filename).split("\n").map {|line| line.chars.map &:to_i }
    # return new(qgrid)
    la = qgrid.size
    lb = qgrid.first.size
    grid = (0...la*5).map { [] }
    (0..4).each do |ma|
      (0..4).each do |mb|
        qgrid.each_with_index do |line,a|
          line.each_with_index do |v,b|
            nv = (v + ma + mb) % 9
            grid[ma*la+a][mb*lb+b] = (nv == 0 ? 9 : nv)
          end
        end
      end
    end
    new(grid)
  end

  def initialize(grid)
    @grid = grid
    @width = grid.first.size
    @height = grid.size
  end

  def cost(start, goal)
    costs = Hash.new {|h,k| h[k] = {} }
    costs[start.first][start.last] = 0
    frontier = [[self.class.heuristic_cost(start,goal),start]]
    count = 0
    while(item = frontier.shift) do
      _,node = item
      x,y = node
      cost_so_far = costs[x][y]
      # puts "Considering #{node} (cost: #{cost_so_far})"
      count += 1
      if node == goal
        puts "Considered #{count}"
        return cost_so_far
      end
      neighbors(node).each do |next_node|
        cost = node_cost(next_node) + cost_so_far
        next unless self.class.new_best(costs,next_node,cost)
        self.class.enqueue(frontier,[cost + self.class.heuristic_cost(next_node, goal), next_node])
      end
      # puts "Frontier #{frontier}" if(count > 10423)
    end
    puts "Found nothing, considered #{count}"
  end

  def node_cost(node)
    x,y = node
    @grid[y][x]
  end

  def neighbors(node)
    x,y = node
    [[0,1],[1,0],[0,-1],[-1,0]].map do |(dx,dy)|
      [x + dx, y + dy]
    end.select do |(x,y)|
      (0...@width).include?(x) && (0...@height).include?(y)
    end
  end

  def self.new_best(costs,node,cost)
    x,y = node
    prev = costs[x][y]
    return false unless prev.nil? || cost < prev
    costs[x][y] = cost
  end

  def self.enqueue(list,item)
    icost,_ = item
    i = list.index {|(xcost,_)| xcost > icost } || list.size
    list.insert(i, item)
  end

  def self.heuristic_cost(a,b)
    (a.first - b.first).abs + (a.last - b.last).abs
  end

end

puts Nav.parse("input/day15").cost([0,0],[499,499])
